addons/mass_mailing/models/mail_notification.py

```diff
@@ -49,53 +40,59 @@ class MailingTrace(models.Model):
+
+    @api.depends('mass_mailing_id.campaign_id')
+    def _compute_campaign_id(self):
+        for notif in self:
+            notif.campaign_id = notif.mass_mailing_id.campaign_id
+
+
+    @api.depends('notification_type', 'mass_mailing_id', 'res_partner_id')
     def _compute_display_name(self):
         for trace in self:
-            trace.display_name = '%s: %s (%s)' % (trace.trace_type, trace.mass_mailing_id.name, trace.id)
-
-    @api.depends('sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled')
-    def _compute_state(self):
-        self.update({'state_update': fields.Datetime.now()})
-        for stat in self:
-            if stat.canceled:
-                stat.state = 'canceled'
-            elif stat.exception:
-                stat.state = 'exception'
-            elif stat.opened or stat.clicked:
-                stat.state = 'opened'
-            elif stat.replied:
-                stat.state = 'replied'
-            elif stat.bounced:
-                stat.state = 'bounced'
-            elif stat.sent:
-                stat.state = 'sent'
-            else:
-                stat.state = 'outgoing'
+            trace.display_name = '%s: %s (%s)' % (trace.notification_type,
+                                                  trace.mass_mailing_id.name or trace.res_partner_id.display_name or '',
+                                                  trace.id)
+
+    def _update_trace_status(self, values):
+        if set(['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled']) & set(values.keys()):
+            values['status_update'] = fields.Datetime.now()
+            if values.get('canceled', False):
+                values['notification_status'] = 'canceled'
+            elif values.get('exception', False):
+                values['notification_status'] = 'exception'
+            elif values.get('opened', False) or values.get('clicked', False):
+                values['notification_status'] = 'opened'
+            elif values.get('replied', False):
+                values['notification_status'] = 'replied'
+            elif values.get('bounced', False):
+                values['notification_status'] = 'bounced'
+            elif values.get('sent', False):
+                values['notification_status'] = 'sent'
 
     @api.model_create_multi
     def create(self, values_list):
         for values in values_list:
-            if 'mail_id' in values:
-                values['mail_id_int'] = values['mail_id']
-        return super(MailingTrace, self).create(values_list)
+            # Change default values if mass mailing is set
+            if 'mass_mailing_id' in values and values.get('mass_mailing_id', False):
+                if 'notification_type' not in values:
+                    values['notification_type'] = 'email'
+                if 'notification_status' not in values:
+                    values['notification_status'] = 'outgoing'
```

Comment:
You set a notification_status here and then modify it again in `_update_trace_status` ?
Seems like this could be factorized.

Reply:
Seems complicate, this condition is to keep the default value (as before) when it acts as a trace.

Reply:
Not really, you would simply need to add at the end of you if/else if case:

```
else:
    # when it acts as a "trace", set 'outgoing' by default
    values['notification_status'] = values.get('notification_status', 'outgoing')
```

I like having everything that deals with a single field at one spot, so you don't have to look in different methods to see with what value you end up with for the create.